services/web/server/tests/unit/with_dbs/02/test_projects_ports_handlers.py

```diff
@@ -0,0 +1,292 @@
+    ],
+)
+async def test_io_workflow(
+    client: TestClient,
+    logged_user: UserInfoDict,
+    user_project: ProjectDict,
+    mock_catalog_service_api_responses: None,
+    expected: type[web.HTTPException],
+):
+    """This tests implements a minimal workflow to support the MVP
+
+    It later stage, this test might be split into smaller unit-tests
+    """
+
+    assert client.app
+
+    project_id = user_project["uuid"]
+
+    # get_project_inputs
+    expected_url = client.app.router["get_project_inputs"].url_for(
+        project_id=project_id
+    )
+    assert URL(f"/v0/projects/{project_id}/inputs") == expected_url
+
+    resp = await client.get(f"/v0/projects/{project_id}/inputs")
+    project_inputs, error = await assert_status(resp, expected_cls=expected)
+
+    if not error:
+        assert project_inputs == {
+            "38a0d401-af4b-4ea7-ab4c-5005c712a546": {
+                "key": "38a0d401-af4b-4ea7-ab4c-5005c712a546",
+                "value": 43,
+                "label": "X",
+            },
+            "fc48252a-9dbb-4e07-bf9a-7af65a18f612": {
+                "key": "fc48252a-9dbb-4e07-bf9a-7af65a18f612",
+                "value": 1,
+                "label": "Z",
+            },
+            "7bf0741f-bae4-410b-b662-fc34b47c27c9": {
+                "key": "7bf0741f-bae4-410b-b662-fc34b47c27c9",
+                "value": False,
+                "label": "on",
+            },
+        }
+
+    # update_project_inputs
+    expected_url = client.app.router["update_project_inputs"].url_for(
+        project_id=project_id
+    )
+    assert URL(f"/v0/projects/{project_id}/inputs") == expected_url
+
+    resp = await client.patch(
+        f"/v0/projects/{project_id}/inputs",
+        json=[{"key": "38a0d401-af4b-4ea7-ab4c-5005c712a546", "value": 42}],
+    )
+    project_inputs, error = await assert_status(resp, expected_cls=expected)
+
+    if not error:
+
+        assert project_inputs == {
+            "38a0d401-af4b-4ea7-ab4c-5005c712a546": {
+                "key": "38a0d401-af4b-4ea7-ab4c-5005c712a546",
+                "value": 42,  # <---- updated
+                "label": "X",
+            },
+            "fc48252a-9dbb-4e07-bf9a-7af65a18f612": {
+                "key": "fc48252a-9dbb-4e07-bf9a-7af65a18f612",
+                "value": 1,
+                "label": "Z",
+            },
+            "7bf0741f-bae4-410b-b662-fc34b47c27c9": {
+                "key": "7bf0741f-bae4-410b-b662-fc34b47c27c9",
+                "value": False,
+                "label": "on",
+            },
+        }
+
+    # get_project_outputs (actual data)
+    expected_url = client.app.router["get_project_outputs"].url_for(
+        project_id=project_id
+    )
+    assert URL(f"/v0/projects/{project_id}/outputs") == expected_url
+
+    resp = await client.get(f"/v0/projects/{project_id}/outputs")
+    project_outputs, error = await assert_status(resp, expected_cls=expected)
+
+    if not error:
```

Comment:
Why would there be an error here? if this is expected in some cases, please provide it as an expectation parameter to the test. I think this allows to better detect issues.

Reply:
this is a technique we use a lot. The test parametrization adds a ``expected`` value with a status code that might be error or not. In the latter case, we perform some extra checks e.g. on the response

the expectation you refer is already checked in ``assert_status( resp, expected_cls=expected) `` . If ``expected`` is a non-error, then that assert function will make sure that error is None